File: 2095-delete-the-middle-node-of-a-linked-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Handle the edge case where the list has only one node.
        if not head.next:
            return None
        
        # Initialize slow and fast pointers.
        # We also need to keep track of the node before the slow pointer (prev).
        slow = head
        fast = head
        prev = None
        
        while fast and fast.next:
            prev = slow
            slow = slow.next
            fast = fast.next.next
            
        # 'slow' is now at the middle, 'prev' is the node before it.
        # Remove the middle node by linking 'prev' to the node after 'slow'.
        prev.next = slow.next
        
        return head

File: 2095-delete-the-middle-node-of-a-linked-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) {
        // If the list has only one node, deleting the middle results in an empty list.
        if (!head || !head->next) {
            return nullptr;
        }

        // Use a dummy node or initialize pointers to track the node before the middle.
        // fast moves twice as fast as slow.
        ListNode* slow = head;
        ListNode* fast = head;
        ListNode* prev = nullptr;

        while (fast && fast->next) {
            prev = slow;
            slow = slow->next;
            fast = fast->next->next;
        }

        // 'prev' is now pointing to the node before the middle node 'slow'.
        // Skip the middle node.
        prev->next = slow->next;
        
        return head;
    }
};